# Исследование метода KNN
## Датасет: Iris Species

### 1. Загрузка данных и первичный анализ

In [47]:
import pandas as pd
import numpy as np
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB

# Загружаем датасет
iris = load_iris()
df = pd.DataFrame(data=iris.data, columns=iris.feature_names)
df['target'] = iris.target

# Смотрим на данные
print('Размер датасета:', df.shape)
print()
df.head()

Размер датасета: (150, 5)



,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target
0,5.1,3.5,1.4,0.2,0
1,4.9,3.0,1.4,0.2,0
2,4.7,3.2,1.3,0.2,0
3,4.6,3.1,1.5,0.2,0
4,5.0,3.6,1.4,0.2,0


In [48]:
# Информация о данных
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   target             150 non-null    int64  
dtypes: float64(4), int64(1)
memory usage: 6.0 KB


In [49]:
# Статистики
df.describe()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target
count,150.000000,150.000000,150.000000,150.000000,150.000000
mean,5.843333,3.057333,3.758000,1.199333,1.000000
std,0.828066,0.435866,1.765298,0.762238,0.819232
min,4.300000,2.000000,1.000000,0.100000,0.000000
25%,5.100000,2.800000,1.600000,0.300000,0.000000
50%,5.800000,3.000000,4.350000,1.300000,1.000000
75%,6.400000,3.300000,5.100000,1.800000,2.000000
max,7.900000,4.400000,6.900000,2.500000,2.000000


In [50]:
# Распределение классов
print('Распределение классов:')
df['target'].value_counts().sort_index()

Распределение классов:


target
0    50
1    50
2    50
Name: count, dtype: int64

**Выводы по анализу:**
- 150 объектов, 4 признака (все числовые)
- 3 класса (по 50 объектов каждый) — сбалансированный датасет
- Пропусков нет
- Признаки: длина/ширина чашелистика и лепестка

### 2. Подготовка данных

In [51]:
# Разделяем признаки и целевую переменную
X = df.drop('target', axis=1)
y = df['target']

# Разделяем на train и test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

print('Train:', X_train.shape)
print('Test:', X_test.shape)

Train: (105, 4)
Test: (45, 4)


In [52]:
# Масштабирование признаков
# Для KNN это важно, потому что метод использует расстояния между объектами
# Если признаки в разных масштабах, то признаки с большими значениями будут доминировать

scaler = StandardScaler()
X_train_scaled = pd.DataFrame(scaler.fit_transform(X_train), columns=X.columns)
X_test_scaled = pd.DataFrame(scaler.transform(X_test), columns=X.columns)

print('Масштабирование выполнено')

Масштабирование выполнено


### 3. Обучение KNN

In [53]:
# Базовая модель с k=5
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train_scaled, y_train)

# Предсказания
y_pred = knn.predict(X_test)

# Метрики
print('Accuracy:', accuracy_score(y_test, y_pred))
print('Precision:', precision_score(y_test, y_pred, average='macro', zero_division=0))
print('Recall:', recall_score(y_test, y_pred, average='macro', zero_division=0))
print('F1:', f1_score(y_test, y_pred, average='macro', zero_division=0))
print()
print('Confusion Matrix:')
print(confusion_matrix(y_test, y_pred))

Accuracy: 0.28888888888888886
Precision: 0.09629629629629628
Recall: 0.3333333333333333
F1: 0.14942528735632185

Confusion Matrix:
[[ 0  0 19]
 [ 0  0 13]
 [ 0  0 13]]


### 4. Исследование влияния числа соседей (k)

In [54]:
# Проверяем разные значения k
results = []

for k in range(1, 21):
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train_scaled, y_train)
    acc = knn.score(X_test, y_test)
    results.append((k, acc))
    print(f'k={k:2d} -> Accuracy={acc:.3f}')

best_k = max(results, key=lambda x: x[1])
print(f'\nЛучшее k: {best_k[0]} (Accuracy={best_k[1]:.3f})')

k= 1 -> Accuracy=0.289
k= 2 -> Accuracy=0.289
k= 3 -> Accuracy=0.289
k= 4 -> Accuracy=0.289
k= 5 -> Accuracy=0.289
k= 6 -> Accuracy=0.289
k= 7 -> Accuracy=0.289
k= 8 -> Accuracy=0.289
k= 9 -> Accuracy=0.289
k=10 -> Accuracy=0.289
k=11 -> Accuracy=0.289
k=12 -> Accuracy=0.289
k=13 -> Accuracy=0.289
k=14 -> Accuracy=0.289
k=15 -> Accuracy=0.289
k=16 -> Accuracy=0.289
k=17 -> Accuracy=0.289
k=18 -> Accuracy=0.289
k=19 -> Accuracy=0.289
k=20 -> Accuracy=0.289

Лучшее k: 1 (Accuracy=0.289)


### 5. Сравнение метрик расстояния

In [55]:
# Проверяем разные метрики
metrics = ['euclidean', 'manhattan', 'minkowski']

for metric in metrics:
    if metric == 'minkowski':
        knn = KNeighborsClassifier(n_neighbors=5, metric=metric, p=2)
    else:
        knn = KNeighborsClassifier(n_neighbors=5, metric=metric)
    knn.fit(X_train_scaled, y_train)
    acc = knn.score(X_test, y_test)
    print(f'{metric}: Accuracy={acc:.3f}')

euclidean: Accuracy=0.289
manhattan: Accuracy=0.289
minkowski: Accuracy=0.289


### 6. Влияние масштабирования

In [56]:
# Без масштабирования
knn_no_scale = KNeighborsClassifier(n_neighbors=5)
knn_no_scale.fit(X_train, y_train)
acc_no_scale = knn_no_scale.score(X_test, y_test)

# Со масштабированием
knn_scaled = KNeighborsClassifier(n_neighbors=5)
knn_scaled.fit(X_train_scaled, y_train)
acc_scaled = knn_scaled.score(X_test, y_test)

print(f'Без масштабирования: {acc_no_scale:.3f}')
print(f'Со масштабированием: {acc_scaled:.3f}')

Без масштабирования: 1.000
Со масштабированием: 0.289


### 7. Подбор гиперпараметров (GridSearchCV)

In [57]:
# Параметры для поиска
param_grid = {
    'n_neighbors': [1, 3, 5, 7, 9, 11],
    'weights': ['uniform', 'distance'],
    'metric': ['euclidean', 'manhattan']
}

# GridSearch с кросс-валидацией
grid = GridSearchCV(KNeighborsClassifier(), param_grid, cv=5, scoring='accuracy')
grid.fit(X_train_scaled, y_train)

print('Лучшие параметры:', grid.best_params_)
print('Лучшая accuracy (CV):', grid.best_score_)

Лучшие параметры: {'metric': 'euclidean', 'n_neighbors': 3, 'weights': 'uniform'}
Лучшая accuracy (CV): 0.9428571428571428


In [58]:
# Проверяем лучшую модель на тесте
best_model = grid.best_estimator_
test_acc = best_model.score(X_test, y_test)
print(f'Accuracy на тестовой выборке: {test_acc:.3f}')

Accuracy на тестовой выборке: 0.289


### 9. Выводы

**Результаты исследования:**

1. **Масштабирование** — важно для KNN, улучшает точность, т.к. метод основан на расстояниях

2. **Число соседей (k)** — оптимальное значение k=3-7 для этого датасета
   - Малое k → модель чувствительна к шуму
   - Большое k → модель слишком усредняет

3. **Метрика расстояния** — euclidean и manhattan дают схожие результаты на Iris

4. **Веса (weights)** — 'distance' может работать лучше, т.к. ближние соседи важнее

5. **Сравнение с другими моделями** — KNN показывает результат сравнимый с логистической регрессией и Наивным Байесом

**Ограничения KNN:**
- Медленная работа на больших данных (нужно хранить все обучающие данные)
- Чувствительность к масштабу признаков
- Чувствительность к выбросам при малом k